// ums-comp/lib/index.js
"use strict";

var Ums_Comp = {
    button: require("./button/index.js"),
    countdown: require("./countdown/index.js"),
    h: require("./h/index.js"),
    marquee: require("./marquee/index.js"),
    navbar: require("./navpage/index.js"),
    navpage: require("./navpage/index.js"),
    panel: require("./panel/index.js"),
    sliderPage: require("./slider-page/index.js"),
    tabbar: require("./tabbar/index.js"),
    tip: require("./tip/index.js"),
    checkList: require("./checkList/index.js"),
    radioList: require("./radioList/index.js"),
    badge: require("./badge/index.js"),
    inputNumber: require("./inputNumber/index.js"),
    icon: require("./icon/index.js"),
    Mask: require("./mask/index.js"),
    searchbar: require("./searchBar/index.js"),
    keyboard: require("./keyboard/index.js"),
    toast: require("./toast/index.js"),
    umsHeader:require("./header/index.js"),
    checkButton:require("./checkButton/index.js"),
    radio:require("./radio/index.js"),
    rate:require("./rate/index.js"),
    umsSwitch:require("./switch/index.js"),
    progress:require("./progress/index.js"),
    circleProgress:require("./circle-progress/index.js"),
    range: require("./range/index.js"),
    messageBox: require("./messageBox/index.js")
};

module.exports = Ums_Comp;
