// ums-comp/packages/ums-toast/src/index.js
'use strict'

/**
 * Created by zmw on 2017/6/12.
 */

var toast,toastConstrucor = Vue.extend(require('./index.vue'))

function Toast(option = {}, callback) {
    let toast
    if (typeof option === 'string') {
        option = {
            message: arguments[0],
            type: ''
        }
        if (typeof callback === 'function') {
            option.callback = callback
        }
    }else if (typeof option === 'object') {
        if (typeof callback === 'function') {
            option.callback = callback
        }
    }
    initToast(option)
};

function initInstance () {
    return new toastConstrucor({
        el: weex.document.createElement('div')
    })
}

function initToast (options) {
    toast = initInstance()
    for (let prop in options) {
        if (options.hasOwnProperty(prop)){
            options[prop] != undefined && (toast[prop] = options[prop])
        }
    }
    toast.close = closeToast
    mountedToast()
}

function mountedToast () {
    toast.$el.setStyle("width", 750, false)
    toast.$el.setStyle("height", 1534, false)
    toast.$nextTick(function () {
        weex.document.body.appendChild(toast.$el)
        toast.animate()
    })
}

function closeToast () {
    toast.callback !== null && toast.callback()
    weex.document.body.removeChild(toast.$el)
}

module.exports = Toast
